src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -32,48 +32,53 @@ protected override void EndProcessing()
         {
             string filePath = null;
             string tempPath = Path.GetTempPath();
-            int attempts = 0;
-            bool temporaryFilePathFound = false;
-            while(attempts++ < 10 && !temporaryFilePathFound)
-            {
-                string fileName = Path.GetRandomFileName();
-                fileName = Path.ChangeExtension(fileName, _extension);
-                filePath = Path.Combine(tempPath, fileName);
-                if(!File.Exists(filePath))
-                {
-                    temporaryFilePathFound = true;
-                }
-            }
-            if(!temporaryFilePathFound)
-            {
-                WriteError(
-                    new ErrorRecord(
-                            new IOException("Could not find an available temporary file name in {tempPath}."),
-                            "NewTemporaryFileResourceUnavailable",
-                            ErrorCategory.ResourceUnavailable,
-                            tempPath));
-                    return;
-            }
 
-            if (ShouldProcess(filePath))
+            if (ShouldProcess(tempPath))
             {
-                try
+                int attempts = 0;
+                bool creationOfFileSuccessful = false;
+
+                // In case the random temporary file already exists, retry 
+                while (attempts++ < 10 && !creationOfFileSuccessful)
                 {
-                    using (new FileStream(filePath, FileMode.CreateNew)) { }
-                    FileInfo file = new FileInfo(filePath);
-                    WriteVerbose("Created temporary file {file.FullName}.");
-                    WriteObject(file);
+                    try
+                    {
+                        string fileName = Path.GetRandomFileName();
+                        fileName = Path.ChangeExtension(fileName, _extension);
+                        filePath = Path.Combine(tempPath, fileName);
+                        using (new FileStream(filePath, FileMode.CreateNew)) { }
+                        creationOfFileSuccessful = true;
+                        WriteVerbose("Created temporary file {filePath}.");
+                    }
+                    catch (IOException) // file already exists -> retry
+                    {
+                        attempts++;
+                    }
+                    catch (Exception exception) // fatal error, which could be e.g. insufficient permissions, etc.
+                    {
+                        WriteError(
+                        new ErrorRecord(
+                                exception,
+                                "NewTemporaryFileWriteError",
+                                ErrorCategory.WriteError,
+                                tempPath));
+                        return;
+                    }
                 }
-                catch (Exception e)
+
+                if (!creationOfFileSuccessful)
                 {
                     WriteError(
```

Comment:
Thanks for the quick review as well. 👍  You have an interesting point. My thinking is: If one has got a script in which a temporary file gets created, then in order to use it, one would need need to use the returned object because otherwise one does not know where the temporary file is. Therefore it is up to the user to check that that their variable is not `$null` and handle it or otherwise the next operation will probably throw a NullReference exception and terminate anyway. But I'll leave this decision of course up to you guys.